hcl/json/structure_test.go

```diff
@@ -301,6 +301,124 @@ func TestBodyPartialContent(t *testing.T) {
 			},
 			1,
 		},
+		{
+			`{"resource": null}`,
+			&hcl.BodySchema{
+				Blocks: []hcl.BlockHeaderSchema{
+					{
+						Type: "resource",
+					},
+				},
+			},
+			&hcl.BodyContent{
+				Attributes: map[string]*hcl.Attribute{},
+				// We don't find any blocks if the value is json null.
+				Blocks: nil,
+				MissingItemRange: hcl.Range{
+					Filename: "test.json",
+					Start:    hcl.Pos{Line: 1, Column: 18, Byte: 17},
+					End:      hcl.Pos{Line: 1, Column: 19, Byte: 18},
+				},
+			},
+			0,
+		},
+		{
+			`{"resource": { "nested": null }}`,
+			&hcl.BodySchema{
+				Blocks: []hcl.BlockHeaderSchema{
+					{
+						Type: "resource",
+					},
```

Comment:
I think this isn't quite hitting the codepath we wanted... the input bytes are fine but HCL is currently understanding it as equivalent to the following native syntax:

```hcl
resource {
  nested = null
}
```

To make HCL expect a level of labels, we need to declare the label names as part of the schema here:

```go
[]hcl.BlockHeaderSchema{
  {
    Type:       "resource",
    LabelNames: []string{"name"},
  }
}
```

The above tells the JSON decoder that it should expect to find a block with a single additional label, which resolves the ambiguity that JSON object syntax is used both for the block attributes and for the block labels.

I think with that extra `LabelNames` setting in the schema, we should expect to see an empty `BodyContent` as the result, because the input here is not a complete definition of a block: it's as if we wrote a native syntax block header without the block body after it:

```hcl
// (not actually valid, because all blocks have bodies in native syntax, but for illustrative purposes...)
resource "nested" /* missing body */
```

I'm assuming that the user's intent with setting a block to null like this is to treat it as if it were not present at all, which can simplify the logic of a program that is generating JSON to be consumed as HCL by letting them not worry about eliding the entire path to the object just to omit the object. In a real-world scenario I'd expect to see this as part of a mixed set/unset map where the generating program is deciding on a per-resource basis whether to include it:

```json
{
  "resource": {
    "a": {
      "some_argument": "foo"
    },
    "b": null,
    "c": {
      "some_argument": "bar"
    }
  }
}
```

...which the JSON decoder should interpret as if the user wrote the following in native syntax:

```hcl
resource "a" {
  some_argument = "foo"
}
# resource "b" not included at all because it is null
resource "c" {
  some_argument = "bar"
}
```

This allowance avoids the need for the generating program to deal with having to omit the property `"b"` altogether when it is unset, continuing our precedent that null is considered the same as not setting the property at all.

Reply:
Yeah - that seems correct, thanks!

I just updated the PR with the extra labels and got the result you'd expect. Thanks 🎉